app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,466 @@
+
+  private int totalFilesForTransfer = -1;
+  private int filesSent = 0;          // Count of number of files transferred until now
+  private ArrayList<FileItem> filesForTransfer = new ArrayList<>();
+
+  private ArrayList<Uri> fileUriArrayList; // For sender device, stores uris of the files
+  private boolean isFileSender = false;    // Whether the device is the file sender or not
+
+  public WifiDirectManager(@NonNull LocalFileTransferActivity activity) {
+    this.activity = activity;
+  }
+
+  /* Initialisations for using the WiFi P2P API */
+  public void createWifiDirectManager(@NonNull SharedPreferenceUtil sharedPreferenceUtil,
+      @NonNull AlertDialogShower alertDialogShower, @Nullable ArrayList<FileItem> filesForTransfer,
+      @Nullable ArrayList<Uri> fileUriArrayList) {
+    this.sharedPreferenceUtil = sharedPreferenceUtil;
+    this.alertDialogShower = alertDialogShower;
+    this.fileUriArrayList = fileUriArrayList;
+    this.filesForTransfer = filesForTransfer;
+    this.isFileSender = (fileUriArrayList != null && fileUriArrayList.size() > 0);
+
+    if(isFileSender) {
+      this.totalFilesForTransfer = fileUriArrayList.size();
+    }
+
+    manager = (WifiP2pManager) activity.getSystemService(Context.WIFI_P2P_SERVICE);
+    channel = manager.initialize(activity, getMainLooper(), null);
+  }
+
+  public void registerWifiDirectBroadcastRecevier() {
+    receiver = new KiwixWifiP2pBroadcastReceiver(manager, channel, this);
+
+    // For specifying broadcasts (of the P2P API) that the module needs to respond to
+    IntentFilter intentFilter = new IntentFilter();
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION);
+
+    activity.registerReceiver(receiver, intentFilter);
+  }
+
+  public void unregisterWifiDirectBroadcastRecevier() {
+    activity.unregisterReceiver(receiver);
+  }
+
+  public void discoverPeerDevices() {
+    manager.discoverPeers(channel, new WifiP2pManager.ActionListener() {
+      @Override
+      public void onSuccess() {
+        displayToast(R.string.discovery_initiated, Toast.LENGTH_SHORT);
+      }
+
+      @Override
+      public void onFailure(int reason) {
+        String errorMessage = getErrorMessage(reason);
+        Log.d(TAG, activity.getString(R.string.discovery_failed) + ": " + errorMessage);
+        displayToast(R.string.discovery_failed, Toast.LENGTH_SHORT);
+      }
+    });
+  }
+
+  @Override
+  public void setWifiP2pEnabled(boolean wifiP2pEnabled) {
+    this.wifiP2pEnabled = wifiP2pEnabled;
+
+    if(wifiP2pEnabled == false) {
+      displayToast(R.string.discovery_needs_wifi, Toast.LENGTH_SHORT);
+      ((Callbacks) activity).clearListOfAvailablePeers();
```

Comment:
You should't be casting the activity to Callbacks every time, cast it in the constructor to a field maybe

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/37c8a468def28a880b16bac25fa55a7ca6454027)